utils/src/main/java/software/amazon/awssdk/utils/async/ByteBufferStoringSubscriber.java

```diff
@@ -0,0 +1,194 @@
+ */
+
+package software.amazon.awssdk.utils.async;
+
+import static software.amazon.awssdk.utils.async.StoringSubscriber.EventType.VALUE;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.utils.Validate;
+import software.amazon.awssdk.utils.async.StoringSubscriber.Event;
+
+/**
+ * An implementation of {@link Subscriber} that stores {@link ByteBuffer} events it receives for retrieval.
+ *
+ * <p>Stored bytes can be read via {@link #transferTo(ByteBuffer)}.
+ */
+@SdkProtectedApi
+public class ByteBufferStoringSubscriber implements Subscriber<ByteBuffer> {
+    /**
+     * The minimum amount of data (in bytes) that should be buffered in memory at a time. The subscriber will request new byte
+     * buffers from upstream until the bytes received equals or exceeds this value.
+     */
+    private final int minimumBytesBuffered;
+
+    /**
+     * The amount of data (in bytes) currently stored in this subscriber. The subscriber will request more data when this value
+     * is below the {@link #minimumBytesBuffered}.
+     */
+    private final AtomicInteger bytesBuffered = new AtomicInteger(0);
+
+    /**
+     * A delegate subscriber that we use to store the buffered bytes in the order they are received.
+     */
+    private final StoringSubscriber<ByteBuffer> storingSubscriber;
+
+    /**
+     * The active subscription. Set when {@link #onSubscribe(Subscription)} is invoked.
+     */
+    private Subscription subscription;
+
+    /**
+     * Create a subscriber that stores at least {@code minimumBytesBuffered} in memory for retrieval.
+     */
+    public ByteBufferStoringSubscriber(int minimumBytesBuffered) {
+        this.minimumBytesBuffered = Validate.isPositive(minimumBytesBuffered, "Data buffer minimum must be positive");
+        this.storingSubscriber = new StoringSubscriber<>(Integer.MAX_VALUE);
+    }
+
+    /**
+     * Transfer the data stored by this subscriber into the provided byte buffer.
+     *
+     * <p>If the data stored by this subscriber exceeds {@code out}'s {@code limit}, then {@code out} will be filled. If the data
+     * stored by this subscriber is less than {@code out}'s {@code limit}, then all stored data will be written to {@code out}.
+     *
+     * <p>If {@link #onError(Throwable)} was called on this subscriber, as much data as is available will be transferred into
+     * {@code out} before the provided exception is thrown (as a {@link RuntimeException}).
+     *
+     * <p>If {@link #onComplete()} was called on this subscriber, as much data as is available will be transferred into
+     * {@code out}, and this will return {@link TransferResult#END_OF_STREAM}.
+     *
+     * <p>Note: This method is NOT THREAD SAFE. It should not be called concurrently, and should not be called by
```

Comment:
Should we make it thread safe? CRT `sendRequestBody` can be invoked multiple times by different threads (not concurrently)  for streaming operations, so we probably need synchronization everywhere anyway.

Reply:
I think I'm wrong in the docs. A call to `transfer` is safe when it happens-before another call to `transfer`. The calls to transfer can come from multiple threads.

In light of that, do you still think we should add synchronization to make sure that no one ever calls `transfer` concurrently? CRT wouldn't do that.

Reply:
I was thinking about the case where thread 1 calls transferTo and while it's still in progress, another thread 2 invokes this method. It is happens-before relation but there may be an overlap. Is it possible?

Reply:
Yeah, that would certainly break, but CRT shouldn’t do it. The previous code makes that same assumption. We could add a synchronized block to the CRT code if you think they might do it? Not sure it’s worth including in the utils class.